Parsing releases.map from cvmfs into xml file

LinaresToine opened the pull request:
Fixes #12419 

#### Status
Tested 

#### Description
Move away frome ReleasesXML in cmssdt. Use a new function to parse `/cvmfs/cms.cern.ch/releases.map` directlyt by the agent to XML format in `/tmp/ReleasesXML`

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
None

#### External dependencies / deployment changes
None

LinaresToine commented:
Hi @hassan11196 apologies for the lack of priority given to this. Here is my proposed solution. Please let me know what you think

amaltaro commented:
@LinaresToine I do not know if you are already expecting for some feedback, but the assumption that cvmfs will always be mounted on the node that is running this call, makes it a hard dependency.
In addition, it looks like you decided to repurpose that module, instead of fetching and parsing the file we have been using for +10 years, it now uses only the raw map file from cvmfs.

I would suggest a different approach: keep the current behavior of that module, but catch a "service unavailable" (or whatever the error was back that day) error and fallback to the cvmfs file. If there is no cvmfs mounted, it fails! otherwise, fetch from the map file and record a line saying this.

hassan11196 commented on src/python/WMCore/Services/TagCollector/TagCollector.py:
```diff
@@ -39,6 +42,71 @@ def __init__(self, url=None, logger=None, configDict=None, **kwargs):
+            if (not anyarch) and ('prodarch=1;' not in line): 
+                continue
+        
+            if production   and production   not in line: 
```
minor nitpick, fix formatting

LinaresToine replied:
@hassan11196 could you please check again? If you mean to remove parentheses for uniformity, it should be ok now. Otherwise, could you please specify?

hassan11196 reviewed (changes requested):
logic looks good, small changes requested

Commit:
Parse cvmfs releasesMap file into xml file

Commit:
Adding unit test

LinaresToine commented:
test this please

LinaresToine commented:
test this please

LinaresToine commented:
test this please

hassan11196 reviewed (approved):
LGTM overall, if you could add the additional check to make sure cvmfs is mounted that might make debugging easier but not a blocker.

amaltaro commented:
Hi @LinaresToine @hassan11196 , apologies for the belated review here.

I didn't really look at the code, but I wanted to make a few remarks for your consideration:
* given the dependency added on `cvmfs`, you might want to start preparing the relevant changes for the CERN and FNAL schedds - otherwise you won't be able to upgrade the agents with the release that provides this feature
* I also think that there should be a graceful check for cvmfs, making a warning record in case it is not available and moving on
* I am not sure I understand those methods that start with `test` or `_test` in the src file. They look like something for the unit tests? If not, then perhaps those methods - and their docstring - could be revisted
* I would suggest to try to resolve the failing unit tests. They might actually be revealing a bug in this implementation
* finally, did you manage to compare the list of ScramArch/CMSSW between the one built with this code and the one provided by CMSSW over HTTP?

LinaresToine commented:
test this please

LinaresToine commented:
Hello @amaltaro 

I have made the following requested changes:
* Added a boolean parameter which checks if cvmfs is mounted before it tries to use it: [TagCollector](https://github.com/dmwm/WMCore/blob/0592f9a947518c6ff83cf56e14426f38c6b1e638/src/python/WMCore/Services/TagCollector/TagCollector.py#L146-L161)
* Moved the test functions to the test module. 

Also, please note that the functionality I am testing takes `releases.map` from cvmfs and creates ReleasesXML. To do this, I create a sample of releases.map with similar format to the original. Such sample is explicit in [TagCollectorTest](https://github.com/dmwm/WMCore/blob/0592f9a947518c6ff83cf56e14426f38c6b1e638/test/python/WMCore_t/Services_t/TagCollector_t/TagCollector_t.py#L123-L129). I ran the tests locally and was able to verify the output `ReleasesXML`:

```
<projects>
  <architecture name="el8_amd64_gcc12">
    <project label="CMSSW_15_0_15_patch3" type="Production" state="Announced" default_micro_arch="x86-64-v3"/>
    <project label="CMSSW_15_0_15_patch4" type="Production" state="Announced" default_micro_arch="x86-64-v3"/>
  </architecture>
  <architecture name="el9_amd64_gcc12">
    <project label="CMSSW_15_0_15_patch3" type="Production" state="Announced" default_micro_arch="x86-64-v3"/>
    <project label="CMSSW_15_0_15_patch4" type="Production" state="Announced" default_micro_arch="x86-64-v3"/>
  </architecture>
</projects>
```

The output from the test matches the expected output and the format is uniform to that in [cmssdt](https://cmssdt.cern.ch/SDT/cgi-bin/ReleasesXML?anytype=1&anyarch=1)

The currently failing tests are independent of the changes introduced.

hassan11196 commented:
Hi @amaltaro @LinaresToine, Thank for the review and repeated changes. lets get this merged so that we can have it in the next WMAgent release, next week hopefully.
